inline_translated.c

```diff
@@ -0,0 +1,554 @@
+    assert(pcpu);
+    if (pcpu->sp >= STACK_CAPACITY-1) {
+        printf("Stack overflow\n");
+        pcpu->state = Cpu_Break;
+        exit_generated_code();
+    }
+    pcpu->stack[++pcpu->sp] = v;
+}
+
+static void inline_translate_program(const Instr_t *prog,
+                           char *out_code, void **entrypoints, int len) {
+    assert(prog);
+    assert(out_code);
+    assert(entrypoints);
+
+    /* An IA-32 instruction "CALL rel32" is used as a trampoline to invoke
+       service routines. A template for it is "call .+0x00000005" */
+    const char call_template_code[] = { 0xe8, 0x00, 0x00, 0x00, 0x00 };
+    const int call_template_size = sizeof(call_template_code);
+
+    int i = 0; /* Address of current guest instruction */
+    char* cur = out_code; /* Where to put new code */
+
+    while (i < len) {
+        decode_t decoded = decode_at_address(prog, i);
+        entrypoints[i] = (void*) cur;
+
+        int addr_exit1 = 0;
+        int addr_exit2 = 0;
+        int addr_exit3 = 0;
+        int addr_puts1 = 0;
+        int addr_puts2 = 0;
+        int addr_push1 = 0;
+        int addr_push2 = 0;
+        int addr_push3 = 0;
+        int addr_abort = 0;
+        int addr_printf = 0;
+        int addr_rand = 0;
+        int addr_str1 = (intptr_t)&str_printf;
+        int addr_str2 = (intptr_t)&str_pop;
+        int addr_str3 = (intptr_t)&str_push;
+        int len = 0;
+        int imm = 0;
+
+        switch(decoded.opcode) {
+        case Instr_Nop:
+            addr_exit1 = (intptr_t)&exit_generated_code - (intptr_t)cur - call_template_size
+                                - (intptr_t)0x2b;
+            len = sizeof(bin_sr_Nop);
+            assert(cur + len - out_code < JIT_CODE_SIZE);
+            memcpy(cur, bin_sr_Nop, len);
+            memcpy(cur+0x2c,&addr_exit1, 4);
+            break;
+        case Instr_Halt:
+            addr_exit1 = (intptr_t)&exit_generated_code - (intptr_t)cur - call_template_size
+                                - (intptr_t)0x11;
+            len = sizeof(bin_sr_Halt);
+            assert(cur + len - out_code < JIT_CODE_SIZE);
+            memcpy(cur, bin_sr_Halt, len);
+            memcpy(cur+0x12,&addr_exit1, 4);
+            break;
+        case Instr_Print:
+            addr_exit1 = (intptr_t)&exit_generated_code - (intptr_t)cur - call_template_size
+                                - (intptr_t)0x57;
+            addr_exit2 = (intptr_t)&exit_generated_code - (intptr_t)cur - call_template_size
+                                - (intptr_t)0x6f;
+            addr_puts1 = (intptr_t)&puts - (intptr_t)cur - call_template_size - (intptr_t)0x61;
+            addr_printf = (intptr_t)&printf - (intptr_t)cur - call_template_size - (intptr_t)0x24;
+            addr_abort = (intptr_t)&abort - (intptr_t)cur - call_template_size - (intptr_t)0x74;
+            len = sizeof(bin_sr_Print);
+            assert(cur + len - out_code < JIT_CODE_SIZE);
+            memcpy(cur, bin_sr_Print, len);

```

Comment:
So you patch the template to make a capsule. I guess this quite heavily depends on compiler, its version and build options.